File: Javascript/Arrays & Hashing/Sum-Of-All-Odd-Length-Subarrays.js
/**
 * @param {number[]} arr
 * @return {number}
 */
const sumOddLengthSubarrays = (arr) => {
  const prefixSumArray = prefixSum(arr);

  let result = 0;
  for (let start = 0; start < prefixSumArray.length; start++) {
    for (let end = start; end < prefixSumArray.length; end += 2) {
      result += sumBetween(start, end, prefixSumArray);
    }
  }

  return result;
};

const prefixSum = (arr) => {
  let prefixSumArray = new Array(arr.length);
  prefixSumArray[0] = arr[0];

  for (let i = 1; i < arr.length; i++) {
    prefixSumArray[i] = prefixSumArray[i - 1] + arr[i];
  }

  return prefixSumArray;
};

const sumBetween = (start, end, prefixSumArr) => {
  return prefixSumArr[end] - (prefixSumArr[start - 1] || 0);
};

File: Rust/Sum-Of-All-Odd-Length-Subarrays.rs
struct Solution;

impl Solution {
    pub fn sum_odd_length_subarrays(arr: Vec<i32>) -> i32 {
        let prefix_sum_array = Self::prefix_sum(&arr);

        let mut result = 0;
        for start in 0..prefix_sum_array.len() {
            let mut end = start;
            while end < prefix_sum_array.len() {
                result += Self::sum_between(start, end, &prefix_sum_array);
                end += 2;
            }
        }
        result
    }

    pub fn prefix_sum(vec: &Vec<i32>) -> Vec<i32> {
        let mut result = vec![0; vec.len()];
        result[0] = vec[0];

        for i in 1..vec.len() {
            result[i] = result[i - 1] + vec[i];
        }

        result
    }

    pub fn sum_between(start: usize, end: usize, prefix_sum_arr: &Vec<i32>) -> i32 {
        prefix_sum_arr[end] - prefix_sum_arr.get(start - 1).unwrap_or(&0)
    }
}
fn main() {}
